core/src/main/java/org/mule/processor/AbstractRequestResponseMessageProcessor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.processor;
+
+import org.mule.DefaultMuleEvent;
+import org.mule.NonBlockingVoidMuleEvent;
+import org.mule.VoidMuleEvent;
+import org.mule.api.CompletionHandler;
+import org.mule.api.MessagingException;
+import org.mule.api.MuleMessage;
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.NonBlockingSupported;
+import org.mule.api.transport.CompletionHandlerReplyToHandlerAdaptor;
+import org.mule.api.transport.ReplyToHandler;
+
+/**
+ * Abstract implementation of a {@link org.mule.api.processor.MessageProcessor} that may performs processing during both
+ * the request and response processing phases while supporting non-blocking execution.
+ */
+public abstract class AbstractRequestResponseMessageProcessor extends AbstractInterceptingMessageProcessor implements
+        NonBlockingSupported
+{
+
+    @Override
+    public final MuleEvent process(MuleEvent event) throws MuleException
+    {
+        if (isNonBlocking(event))
+        {
+            return processNonBlocking(event);
+        }
+        else
+        {
+            return processBlocking(event);
+        }
+    }
+
+    protected MuleEvent processBlocking(MuleEvent event) throws MuleException
+    {
+        MessagingException exception = null;
+        try
+        {
+            return processResponse(processNext(processRequest(event)));
+        }
+        catch (MessagingException e)
+        {
+            exception = e;
+            throw e;
+        }
+        finally
+        {
+            processFinallly(event, exception);
```

Comment:
Then it's propagated up, the same as if you threw an exception inside a finally block. That said, processFinally doesn't declare an exception, so it'd only be runtime exceptions.  See any issue with this?